.changeset/hmr.md

```diff
@@ -0,0 +1,20 @@
+---
+"@remix-run/dev": minor
+"@remix-run/react": minor
+"@remix-run/server-runtime": minor
+---
+
+Hot Module Replacement and Hot Data Revalidation
+
+- Requires `unstable_dev` future flag to be enabled
+- HMR provided through React Refresh
+
+Features:
+- HMR for component and style changes
+- HDR when loaders for current route change
+
+Known limitations for MVP:
+- Only implemented for React via React Refresh
+- No `import.meta.hot` API exposed yet
+- Revalidates _all_ loaders on route when loader changes are detected
+- Loader changes do not account for imported dependencies changing
```

Comment:
Which of these are short-term limitations and which are really hard problems that will have to be solved longer-term (or never solved at all)?

Reply:
> - Revalidates _all_ loaders on route when loader changes are detected
> - Loader changes do not account for imported dependencies changing

👆 short-term, should be resolved before HMR is stabilized

> - No `import.meta.hot` API exposed yet

👆 we already prototyped a `import.meta.hot` API so my gut is that we can also do this short-term as needed

> - Only implemented for React via React Refresh

👆 If we have `import.meta.hot` API, then HMR should be implementable in user-land for any other rendering frameworks/libraries. Out-of-the-box HMR support for other renderers would come after we support that renderer (e.g. after `@remix-run/vue` exists or something like that).

Reply:
Sweet. Thanks!

Reply:
Also, I've noticed that HMR can be quite slow (several seconds after save). Is that expected? If you want to play around with a real world project that has this issue, check: https://github.com/epicweb-dev/rocket-rental

Reply:
I don't think my server is restarting. I'm doing the purge require cache approach instead

Reply:
> I don't think my server is restarting. I'm doing the purge require cache approach instead

Ah whoops. Saw `tsx watch` and made assumptions, but you are only watching for changes in `./index.js` and not `build/index.js`. Make sense.

Yea so to summarize, we can optimize HMR speed directly, but have limitations on HDR speed depending on how fast your app server responds with an up-to-date build hash via `__REMIX_ASSETS_MANIFEST` endpoint. Should be able to at least speed up HMR part as part of stabilizing this feature.

Reply:
That's great to hear! Thanks Pedro!